### ISSUE TITLE
Use newer build-system on Windows

### ISSUE THREAD

**Issue by masatake** [universal-ctags/ctags#3805]
@k-takata, @leleilu008, and @edam, how we should do?

              `cl : Command line warning D9002 : ignoring unknown option '/experimental:preprocessor'`

I think the compiler in AppVeyor is too old @masatake! 😆 This should fix the issue. (Actually, in later compilers, the option is no longer experimental: `/Zc:preprocessor`). The build issue is that the preprocessor incorrectly expands `__VA_ARGS__` in macros when those args have come from `...`. I can reproduced it on windows (in a VM) and this fixes it with a later MSVC compiler.

Can we update the AppVeyor compiler? Or is that not really possible?

_Originally posted by @edam in https://github.com/universal-ctags/ctags/issues/3795#issuecomment-1684897069_

**Comment by k-takata** [comment]
C99 support in Visual C++ was improved in VC++ 2015, so dropping 2013 might be better for future development.
VC++ 2013 was not old at the start of this project, but it is quite old now.
Should we define a policy on which versions of VC++ are to be supported? (E.g. two or three major versions?)

**Comment by edam** [comment]
The non-conformant preprocessor issue in #3795 is fixed with the "conformant" option. (Why msvc needs a "conformant" option, is a different question... 🤷) And I believe that option is only available from VS2019 onwards. (It is an experimental option in VS2019, and a regular non-experimental feature after that version.) So my vote would be for a newer version to be supported. 😄

**Comment by leleliu008** [comment]
I doubt that if there is really someone tries to build ctags with Visual Studio, because it's not full-featured, our daily builds for Windows are full-featured only. https://github.com/universal-ctags/ctags-win32

Due to the above reason, I think we only need to support the lasted two releases of Visual Studio, that is `Visual Studio 2022` and `Visual Studio 2019` at this point. That is the versions GitHub Actions supported.

**Comment by leleliu008** [comment]
Is it necessary to maintain the `msbuild` system? `nmake` can't meet needs? `win32/ctags_vs2013.vcxproj` use `Visual Studio 2013`

**Comment by leleliu008** [comment]
I tried to build ctags with `Visual Studio 2022 / MSBuild`, I got following error message:

```
  Building solution configuration "Release|Win32".
Project "D:\a\test\test\win32\ctags_vs2013.sln" (1) is building "D:\a\test\test\win32\ctags_vs2013.vcxproj" (2) on node 1 (default targets).
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(456,5): error MSB8020: The build tools for Visual Studio 2013 (Platform Toolset = 'v120') cannot be found. To build using the v120 build tools, please install Visual Studio 2013 build tools.  Alternatively, you may upgrade to the current Visual Studio tools by selecting the Project menu or right-click the solution, and then selecting "Retarget solution". [D:\a\test\test\win32\ctags_vs2013.vcxproj]
```

I don't known the advantage of `MSBuild`, I wonder why we support both `nmake` and `msbuild`? If we want to support `msbuild, looks like `win32/ctags_vs2015.vcxproj` `win32/ctags_vs2017.vcxproj` `win32/ctags_vs2019.vcxproj` `win32/ctags_vs2022.vcxproj` also need to be included?

**Comment by masatake** [comment]
@k-takata wrote:
> Should we define a policy on which versions of VC++ are to be supported? (E.g. two or three major versions?)

I don't have any comments about this. If we, the maintainers, are comfortable, we should do so.
Writing the policy explicitly is nice.

**Comment by k-takata** [comment]
OK, supporting the latest two major versions of MSVC might be enough.

> Is it necessary to maintain the `msbuild` system?

Maybe no? It is necessary if someone wants to build ctags on Visual Studio IDE. Visual Studio IDE is especially useful when using its debugger. I think it is better than GDB (at least on Windows). But, I rarely wanted to use Visual Studio's debugger, because printf-debugging was almost enough for debugging ctags.

**Comment by edam** [comment]
Just checking to see if a decision was reached here?

VS2019+ support is needed for the new parser in #3795 (they added a "conformant" mode for the preprocessor in VS2019! 🤯)

**Comment by masatake** [comment]
In my understanding, #3815 may solve this.

**Comment by masatake** [comment]
Closed via #3815.
